valves/__init__.py

```diff
@@ -0,0 +1,105 @@
+
+try:
+    import pandas as pd
+
+    _PANDAS_AVAILABLE = True
+except ImportError:
+    _PANDAS_AVAILABLE = False
+
+
+try:
+    import dask.dataframe as dd
+
+    _DASK_AVAILABLE = True
+except ImportError:
+    _DASK_AVAILABLE = False
+
+
+def _raise_dataf_error(dataf):
+    """Raises error based on availability of dataframe package."""
+    installed_packages = []
+    if _DASK_AVAILABLE:
+        installed_packages.append("dask")
+    if _PANDAS_AVAILABLE:
+        installed_packages.append("pandas")
+    if _POLARS_AVAILABLE:
+        installed_packages.append("polars")
+    raise ValueError(
+        f"Recieved {type(dataf)} dataframe. Are you sure this is installed? We can confirm {installed_packages} are installed."
+    )
+
+
+def sessionize(dataf, user_col="user", ts_col="timestamp", threshold=20 * 60):
+    """
+    Adds a session to the dataset, using multiple dispatch to proxy the correct backend.
+
+    This function is meant to be used in a `.pipe()`-line.
+
+    Arguments:
+        - dataf: pandas, polars or dask dataframe
+        - user_col: name of the column containing the user id
+        - ts_col: name of the column containing the timestamp
+        - threshold: time in seconds to consider a user inactive
+    """
+    if _DASK_AVAILABLE and isinstance(dataf, dd.DataFrame):
+        # Import dask-related stuff only if dask is detected
+        from .dask import sessionize as sess_dd
+
+        return sess_dd(dataf, user_col, ts_col, threshold)
+    if _PANDAS_AVAILABLE and isinstance(dataf, pd.DataFrame):
+        # Import pandas-related stuff only if pandas is detected
+        from .pandas import sessionize as sess_pd
+
+        return sess_pd(dataf, user_col, ts_col, threshold)
+    if _POLARS_AVAILABLE and isinstance(dataf, pl.DataFrame):
+        # Import polars-related stuff only if polars is detected
+        from .polars import sessionize as sess_pl
+
+        return sess_pl(dataf, user_col, ts_col, threshold)
+    _raise_dataf_error(dataf)
+
+
+def bayes_average(
+    dataf, group_cols, target_col, C, prior_mean=None, out_col="bayes_avg"
+):
+    r"""
+    Computes the Bayes average for a target column.
+
+    The average is calculated per formula found on [wikipedia](https://en.wikipedia.org/wiki/Bayesian_average).
+
+    $$
+    \bar{x}=\frac{C m+\sum_{i=1}^{n} x_{i}}{C+n}
+    $$
+
+    This function is meant to be used in a `.pipe()`-line.
+
+    Arguments:
+        - dataf: dask dataframe
+        - group_cols: list of columns to group by
+        - target_col: name of the column containing the target value, typically a rating
+        - C: smoothing parameter
+        - prior_mean: optional, a prior mean to use instead of the mean of the target column
+        - out_col: name of the column to output
+    """
+    if _DASK_AVAILABLE and isinstance(dataf, dd.DataFrame):
+        # Import dask-related stuff only if dask is detected
+        from .dask import bayes_average as bayes_dd
```

Comment:
Not a blocker for this PR, but we could already import it under the import branches. If `_DASK_AVAILABLE` we already know that `bayes_average ` is imported as `bayes_dd`.

For now there is only one import, but if we grow with imports, I think that will keep things a bit more clean.

Reply:
It sure took me a while, but it seems that I finally found a use-case for `try-except-else`! 

Will make the edit before I merge.